BUG: signal.medfilt and ndimage.median_filter not returning correct output

### Describe your issue.

After investigating a regression with some of the tests at work after upgrading from Scipy 1.14.1 to 1.15.1 I noticed some odd behavior in Scipy 1.15.1.

For certain inputs, scipy.signal.medfilt and scipy.ndimage.median_filter will not always return the correct input (kinda random).

For example, if I try to median filter the following sequence with a kernel size 9 and zero padding: `[272, 58, 67, 163, 463, 608, 87, 108, 1378]`
I sometimes get this wrong result `[58, 67, 87, 108,  *272*, 108, 108, 108, 87]`
instead of this correct result: `[ 58, 67, 87, 108, *163*, 108, 108, 108,  87]`
This example can take multiple runs to reproduce.

I have another example though that can somewhat reliably reproduce this issue. See the reproducing code example below

### Reproducing Code Example

```python
import numpy as np
from scipy.signal import medfilt
from scipy.ndimage import median_filter


# An input that can somewhat reliably reproduce the bug in a short amount of iterations
# I also managed to replicate this issue for x = [272, 58, 67, 163, 463, 608, 87, 108, 1378] with kernel size 9 though
# I had to rerun the script multiple times for the wrong result to appear. In all cases of wrong results, I got
# [58, 67, 87, 108,  *272*, 108, 108, 108, 87] instead of [58, 67, 87, 108, *163*, 108, 108, 108, 87]
x = [272, 58, 67, 163, 463, 608, 87, 108, 1378, 72, 136, 566, 747, 122, 249, 302, 48, 181, 60, 104, 119, 206, 349, 223, 58, 105, 116, 136, 114, 386, 108, 112, 127, 91, 120, 110, 117, 117, 63, 65, 128, 97, 96, 99, 124, 107, 105, 122, 121, 129, 77, 104, 141, 132, 216, 301, 130, 113, 92, 89, 99, 311, 405, 41, 89, 103, 5044, 116, 71, 96, 92, 123, 4561, 84]

# Expected result after median filtering with kernel size 9 (zero-padded edges), computed using Scipy 1.14.1
# Especially note that the result from Scipy 1.15.1 is often wrong in the fifth element (y[4] != expected[4]),
# with value 272 instead of 163
expected = [58, 67, 87, 108, 163, 108, 136, 163, 463, 136, 136, 249, 249, 181, 181, 181, 122, 122, 181, 181, 119, 119, 116, 119, 119, 136, 116, 114, 114, 114, 116, 114, 114, 117, 112, 112, 117, 110, 110, 99, 99, 99, 99, 105, 107, 107, 107, 107, 121, 121, 122, 129, 130, 130, 130, 130, 130, 130, 130, 113, 99, 99, 99, 103, 103, 103, 96, 96, 103, 103, 96, 92, 84, 84]


for idx in range(1_000):
    np.testing.assert_array_equal(medfilt(x, 9), expected)

    # uncomment this line and comment line above to test scipy.ndimage.median_filter
    #np.testing.assert_array_equal(median_filter(x, 9, mode="constant"), expected)
```

### Error message

```shell
For medfilt:

Traceback (most recent call last):
  File "/home/kristofer/worktrees/develop/bug_demo.py", line 19, in <module>
    np.testing.assert_array_equal(medfilt(x, 9), expected)
  File "/home/kristofer/worktrees/develop/dist/export/python/virtualenvs/python-default/3.12.3/lib/python3.12/site-packages/numpy/_utils/__init__.py", line 85, in wrapper
    return fun(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/kristofer/worktrees/develop/dist/export/python/virtualenvs/python-default/3.12.3/lib/python3.12/site-packages/numpy/testing/_private/utils.py", line 1020, in assert_array_equal
    assert_array_compare(operator.__eq__, actual, desired, err_msg=err_msg,
  File "/home/kristofer/worktrees/develop/dist/export/python/virtualenvs/python-default/3.12.3/lib/python3.12/site-packages/numpy/testing/_private/utils.py", line 884, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

Mismatched elements: 1 / 74 (1.35%)
Max absolute difference among violations: 109
Max relative difference among violations: 0.66871166
 ACTUAL: array([ 58,  67,  87, 108, 272, 108, 136, 163, 463, 136, 136, 249, 249,
       181, 181, 181, 122, 122, 181, 181, 119, 119, 116, 119, 119, 136,
       116, 114, 114, 114, 116, 114, 114, 117, 112, 112, 117, 110, 110,...
 DESIRED: array([ 58,  67,  87, 108, 163, 108, 136, 163, 463, 136, 136, 249, 249,
       181, 181, 181, 122, 122, 181, 181, 119, 119, 116, 119, 119, 136,
       116, 114, 114, 114, 116, 114, 114, 117, 112, 112, 117, 110, 110,...


For median_filter:
Traceback (most recent call last):
  File "/home/kristofer/worktrees/develop/bug_demo.py", line 22, in <module>
    np.testing.assert_array_equal(median_filter(x, 9, mode="constant"), expected)
  File "/home/kristofer/worktrees/develop/dist/export/python/virtualenvs/python-default/3.12.3/lib/python3.12/site-packages/numpy/_utils/__init__.py", line 85, in wrapper
    return fun(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/kristofer/worktrees/develop/dist/export/python/virtualenvs/python-default/3.12.3/lib/python3.12/site-packages/numpy/testing/_private/utils.py", line 1020, in assert_array_equal
    assert_array_compare(operator.__eq__, actual, desired, err_msg=err_msg,
  File "/home/kristofer/worktrees/develop/dist/export/python/virtualenvs/python-default/3.12.3/lib/python3.12/site-packages/numpy/testing/_private/utils.py", line 884, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

Mismatched elements: 1 / 74 (1.35%)
Max absolute difference among violations: 109
Max relative difference among violations: 0.66871166
 ACTUAL: array([ 58,  67,  87, 108, 272, 108, 136, 163, 463, 136, 136, 249, 249,
       181, 181, 181, 122, 122, 181, 181, 119, 119, 116, 119, 119, 136,
       116, 114, 114, 114, 116, 114, 114, 117, 112, 112, 117, 110, 110,...
 DESIRED: array([ 58,  67,  87, 108, 163, 108, 136, 163, 463, 136, 136, 249, 249,
       181, 181, 181, 122, 122, 181, 181, 119, 119, 116, 119, 119, 136,
       116, 114, 114, 114, 116, 114, 114, 117, 112, 112, 117, 110, 110,...
```

### SciPy/NumPy/Python version and system information

```shell
I managed to reproduce this issue on Windows 11 (not WSL) using Python 3.13, Scipy 15.1, and Numpy 2.21.

I also managed to reproduce this issue using Python 3.12 on Ubuntu 24.04 running in WSL:
>>> import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info); scipy.show_config()
1.15.1 2.2.1 sys.version_info(major=3, minor=12, micro=3, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.12.7/lib/python3.12/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.12.7/lib/python3.12/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.28 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
    pc file directory: /project
    version: 0.3.28
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.12.7/lib/python3.12/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.12.7/lib/python3.12/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.28 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
    pc file directory: /project
    version: 0.3.28
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.13.6
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-p8gqjxwf/overlay/lib/python3.12/site-packages/pythran
    version: 0.17.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp312-cp312/bin/python
  version: '3.12'
```

--- a/scipy/ndimage/src/_rank_filter_1d.cpp
+++ b/scipy/ndimage/src/_rank_filter_1d.cpp
@@ -155,6 +155,7 @@ void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
                   int mode, T cval, int origin) {
   int i, arr_len_thresh, lim = (win_len - 1) / 2 - origin;
   int lim2 = arr_len - lim;
+  int offset;
   Mediator *m = MediatorNew(win_len, rank);
   T *data = new T[win_len];
 
@@ -180,7 +181,13 @@ void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
     }
     break;
   case WRAP:
-    for (i = arr_len - lim - 1 - 2 * origin; i < arr_len; i++) {
+    if (win_len % 2 == 0) {
+        offset = 2;
+    }
+    else {
+        offset = 0;
+    }
+    for (i = arr_len - lim - offset - 2 * origin; i < arr_len; i++) {
       MediatorInsert(data, m, in_arr[i]);
     }
     break;
@@ -222,7 +229,7 @@ void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
     }
     break;
   case WRAP:
-    for (i = 0; i < win_len; i++) {
+    for (i = 0; i < lim; i++) {
       MediatorInsert(data, m, in_arr[i]);
       out_arr[lim2 + i] = data[m->heap[0]];
     }

--- a/scipy/ndimage/src/_rank_filter_1d.cpp
+++ b/scipy/ndimage/src/_rank_filter_1d.cpp
@@ -158,6 +158,10 @@ void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
   int offset;
   Mediator *m = MediatorNew(win_len, rank);
   T *data = new T[win_len];
+  for (int i = 0; i < win_len; ++i) {
+    data[i] = 0;
+  }
+
 
   switch (mode) {
   case REFLECT:

--- a/scipy/ndimage/tests/test_filters.py
+++ b/scipy/ndimage/tests/test_filters.py
@@ -5,7 +5,7 @@ import re
 
 import numpy as np
 import pytest
-from numpy.testing import suppress_warnings
+from numpy.testing import suppress_warnings, assert_allclose
 from pytest import raises as assert_raises
 from scipy import ndimage
 from scipy._lib._array_api import (
@@ -2812,3 +2812,42 @@ def test_byte_order_median(xp):
     b = xp.arange(9, dtype='>f4').reshape(3, 3)
     t = ndimage.median_filter(b, (3, 3))
     assert_array_almost_equal(ref, t)
+
+
+
+@pytest.mark.parametrize("filter_size, exp", [
+    # expected results from SciPy 1.14.1
+    (20, 0.25754605),
+    (10,
+     np.array([0.25266576, 0.27894721, 0.30445588, 0.30958242, 0.30445588, 0.30445588,
+               0.27894721, 0.30445588, 0.27894721, 0.30445588, 0.30445588, 0.25754605,
+               0.22183391, 0.18015438, 0.22183391, 0.25266576, 0.25754605, 0.25266576,
+               0.25266576, 0.25266576]),
+     ),
+     # a median filter size of 1 is just an identity operation
+     (1,
+      np.array([0.30958242, 0.17555138, 0.34343917, 0.27894721, 0.03767094, 0.39024894,
+                0.30445588, 0.31442572, 0.05124545, 0.18015438, 0.14831921, 0.370706,
+                0.25754605, 0.32910465, 0.17736568, 0.09089549, 0.22183391, 0.0255269,
+                0.33105247, 0.25266576]),
+     ),
+     # testing odd-sized filters >1 makes sense too
+     (3,
+      np.array([0.25266576, 0.30958242, 0.27894721, 0.27894721, 0.27894721, 0.30445588,
+                0.31442572, 0.30445588, 0.18015438, 0.14831921, 0.18015438, 0.25754605,
+                0.32910465, 0.25754605, 0.17736568, 0.17736568, 0.09089549, 0.22183391,
+                0.25266576, 0.30958242]),
+     ),
+     (15,
+      np.array([0.27894721, 0.25266576, 0.25266576, 0.25266576, 0.27894721, 0.27894721,
+                0.27894721, 0.27894721, 0.25754605, 0.25754605, 0.22183391, 0.22183391,
+                0.25266576, 0.25266576, 0.22183391, 0.22183391, 0.25266576, 0.25266576,
+                0.25754605, 0.25754605]),
+     ),
+])
+def test_gh_22250(filter_size, exp):
+    rng = np.random.default_rng(42)
+    image = np.zeros((20,))
+    noisy_image = image + 0.4 * rng.random(image.shape)
+    result = ndimage.median_filter(noisy_image, size=filter_size, mode='wrap')
+    assert_allclose(result, exp)

--- a/scipy/ndimage/tests/test_filters.py
+++ b/scipy/ndimage/tests/test_filters.py
@@ -5,7 +5,7 @@ import re
 
 import numpy as np
 import pytest
-from numpy.testing import suppress_warnings, assert_allclose
+from numpy.testing import suppress_warnings, assert_allclose, assert_array_equal
 from pytest import raises as assert_raises
 from scipy import ndimage
 from scipy._lib._array_api import (
@@ -2814,7 +2814,6 @@ def test_byte_order_median(xp):
     assert_array_almost_equal(ref, t)
 
 
-
 @pytest.mark.parametrize("filter_size, exp", [
     # expected results from SciPy 1.14.1
     (20, 0.25754605),
@@ -2851,3 +2850,10 @@ def test_gh_22250(filter_size, exp):
     noisy_image = image + 0.4 * rng.random(image.shape)
     result = ndimage.median_filter(noisy_image, size=filter_size, mode='wrap')
     assert_allclose(result, exp)
+
+
+def test_gh_22333():
+    x = np.array([272, 58, 67, 163, 463, 608, 87, 108, 1378])
+    expected = [58, 67, 87, 108, 163, 108, 108, 108, 87]
+    actual = ndimage.median_filter(x, size=9, mode='constant')
+    assert_array_equal(actual, expected)

